dev/src/bulk-writer.ts

```diff
@@ -0,0 +1,428 @@
+   *   console.log('Successfully executed all writes');
+   * });
+   */
+  async waitForPendingWrites(): Promise<void> {
+    const trackedBatches = this.batchQueue;
+    this.sendBatches();
+    const writePromises = trackedBatches.map(batch => {
+      batch.pending = true;
+      return this.resultsMap.get(batch.tag)!.promise;
+    });
+    await Promise.all(writePromises);
+  }
+
+  /**
+   * Return the first eligible batch that can hold a write to the provided
+   * reference, or creates one if no eligible batches are found.
+   */
+  private getEligibleBatch(ref: DocumentReference): BulkCommitBatch {
+    let batch = null;
+    let logWarning = true;
+    for (const bulkBatch of this.batchQueue) {
+      if (bulkBatch.refsInBatch.has(ref)) {
+        logWarning = false;
+      }
+
+      if (
+        bulkBatch.pending === false &&
+        bulkBatch.writeBatch.opCount < BULK_WRITER_MAX_BATCH_SIZE &&
+        !bulkBatch.refsInBatch.has(ref)
+      ) {
+        batch = bulkBatch;
+      }
+    }
+
+    if (batch === null) {
+      if (logWarning) {
+        logger(
+          'BulkWriter',
+          null,
+          'Writing to the same document multiple times will slow down BulkWriter. ' +
+            'Write to unique documents in order to maximize throughput.'
+        );
+      }
+      return this.createNewBatch();
+    } else {
+      return batch;
+    }
+  }
+
+  /**
+   * Create a new batch on the BatchQueue and a mapping from the batch to its
+   * result.
+   */
+  private createNewBatch(): BulkCommitBatch {
+    const tag = requestTag();
+    const deferred = new Deferred<BatchWriteResult[]>();
+    const newBatch = {
+      writeBatch: this.firestore.batch(),
+      refsInBatch: new Set(),
+      tag,
+      pending: false,
+      result: deferred,
+    } as BulkCommitBatch;
+
+    this.batchQueue.push(newBatch);
+    this.resultsMap.set(tag, deferred);
+    return newBatch;
+  }
+
+  /**
+   * Return the result of the write at the given batch tag and operation count.
+   */
+  private getResult(tag: string, opCount: number): Promise<WriteResult> {
+    assert(this.resultsMap.has(tag), 'requestTag not found');
+    return this.resultsMap.get(tag)!.promise.then(results => {
+      assert(
+        results.length >= opCount,
+        'opCount invalid' + results.length + opCount
+      );
+      const result = results[opCount - 1];
+      if (result.writeTime) {
+        return new WriteResult(result.writeTime);
+      } else {
+        throw result.status;
+      }
+    });
+  }
+
+  /**
+   * Attempts to send batches starting from the front of the BatchQueue until a
+   * batch cannot be sent.
+   *
+   * After a batch is complete, try sending batches again.
+   */
+  private sendBatches(): void {
+    let counter = 0;
+    for (const batch of this.batchQueue) {
+      // Only send up until the first batch that cannot be sent.
+      if (!this.isBatchSendable(batch)) {
+        break;
```

Comment:
Could this be `continue`?

Reply:
When reading this I had the same pause @schmidt-sebastian had. I think there are more idiomatic ways to express this idea (i.e. process the first N elements that meet a certain condition).  This would make more sense as a `while()` loop or perhaps a combination of `Array.prototype.findIndexOf()` and `Array.prototype.slice()`.  Since you ultimately want to return the elements processed I'd suggest with the latter. Plus it lets you get rid of the `counter` variable entirely.

Reply:
Ended up using a `while()` loop since I am removing the batch from the BatchQueue after the batch is committed instead of using `slice()` to remove the Batches are the end of the loop.